src/auxiliary/mod.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright 2018 OpenST Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//!This module is about interaction with auxiliary block chain.
+
+use futures::Future;
+use web3::contract::{Contract, Options};
+use web3::transports::Http;
+use web3::types::{H160, Address, H256};
+use rlp;
+use sha3::{Digest, Keccak256};
+
+use super::ethereum::Ethereum;
+use ethereum::types::{Block, Error, ErrorKind};
+
+const REPORT_BLOCK_ESTIMATED_GAS: i32 = 3_000_000;
+
+/// Reports block on block store if not already reported.
+///
+/// # Arguments
+///
+/// * `blockchain` - A blockchain object.
+/// * `event_loop` - The reactor's event loop to handle the tasks spawned.
+/// * `block_store_address` - The address of origin block store.
+/// * `validator_address` - The address of origin validator address.
+pub fn report_block(
+    block_chain: &Ethereum,
+    event_loop: &tokio_core::reactor::Handle,
+    block_store_address: Address,
+    validator_address: Address,
+    block: &Block,
+) {
+    let encoded_block = rlp::encode(block);
+
+    let block_hash = H256::from(Keccak256::digest(encoded_block.as_slice()).as_slice());
+
+    let contract: Contract<Http> = block_chain.contract_instance(
+        block_store_address,
+        include_bytes!( "../contract/abi/BlockStore.json"),
+    );
+    let event_loop_clone = event_loop.clone();
+
+    let call_future = contract.query("isBlockReported", block_hash, H160::from(validator_address), Options::default(), None)
+        .then(move |result: Result<bool, web3::contract::Error>| {
+            let block_reported = match result {
+                Ok(is_reported) => {
+                    if !is_reported {
+                        let report_future = contract
+                            .call("reportBlock", encoded_block, H160::from(validator_address), Options::with(|opt| {
+                                opt.gas = Some(REPORT_BLOCK_ESTIMATED_GAS.into())
+                            }))
+                            .then(move |tx| {
+                                println!("Block reported got tx: {:?}", tx);
+                                Ok(())
+                            });
+                        event_loop_clone.spawn(report_future);
+                    }
+                    Ok(!is_reported)
```

Comment:
What purpose does this serve?

Reply:
Status whether the block is reported or not is logged.

Reply:
Ah, whether *this code* did a transaction to report the block. Maybe make it more readable by improving the naming or documentation or both.